pkg/config/usagereportingconfig.go

```diff
@@ -0,0 +1,265 @@
+			if _, err := cronexpr.Parse(val); err != nil {
+				log.Tracef("Could not use %s (%s) it is not a proper cron schedule", qaUsageReportingOfflineScheduleEnvVar, val)
+			} else {
+				log.Tracef("Using %s (%s) rather than the default (%s) for non-QA", qaUsageReportingOfflineScheduleEnvVar, val, u.reportSchedule)
+				u.reportSchedule = val
+				u.qaVars = true
+			}
+		}
+
+		// Check the cron expressions
+		cron, err := cronexpr.Parse(u.Schedule)
+		if err != nil {
+			exception.Throw(ErrBadConfig.FormatError(pathUsageReportingSchedule))
+		}
+		nextTwoRuns := cron.NextN(time.Now(), 2)
+		if len(nextTwoRuns) != 2 {
+			exception.Throw(ErrBadConfig.FormatError(pathUsageReportingSchedule))
+		}
+		u.reportGranularity = int(nextTwoRuns[1].Sub(nextTwoRuns[0]).Milliseconds())
+
+		// if no QA env vars are set then validate the schedule is at least hourly
+		if nextTwoRuns[1].Sub(nextTwoRuns[0]) < time.Hour && !u.qaVars {
+			exception.Throw(ErrBadConfig.FormatError(pathUsageReportingSchedule))
+		}
+	}
+}
+
+// GetURL - Returns the usage reporting URL
+func (u *UsageReportingConfiguration) GetURL() string {
+	return u.URL
+}
+
+// CanPublishUsage - Returns the publish boolean
+func (u *UsageReportingConfiguration) CanPublishUsage() bool {
+	return u.Publish
+}
+
+// CanPublishMetric - Returns the publish metric boolean
+func (u *UsageReportingConfiguration) CanPublishMetric() bool {
+	return u.PublishMetric
+}
+
+// GetInterval - Returns the publish interval
+func (u *UsageReportingConfiguration) GetInterval() time.Duration {
+	return u.Interval
+}
+
+// IsOfflineMode - Returns the offline boolean
+func (u *UsageReportingConfiguration) IsOfflineMode() bool {
+	return u.Offline
+}
+
+// GetSchedule - Returns the schedule string
+func (u *UsageReportingConfiguration) GetSchedule() string {
+	return u.Schedule
+}
+
+// GetReportSchedule - Returns the offline schedule string for creating reports
+func (u *UsageReportingConfiguration) GetReportSchedule() string {
+	return u.reportSchedule
+}
+
+// GetReportGranularity - Returns the granularity used in the offline reports
+func (u *UsageReportingConfiguration) GetReportGranularity() int {
+	return u.reportGranularity
+}
+
+// UsingQAVars - Returns the offline boolean
+func (u *UsageReportingConfiguration) UsingQAVars() bool {
+	return u.qaVars
+}
+
+// AddUsageReportingProperties - Adds the command properties needed for Uage Reporting Settings
+func AddUsageReportingProperties(props properties.Properties) {
+	props.AddBoolProperty(pathUsageReportingPublish, true, "Indicates if the agent can publish usage event to Amplify platform. Default to true")
+	props.AddBoolProperty(pathUsageReportingPublishMetric, false, "Indicates if the agent can publish metric event to Amplify platform. Default to false")
+	props.AddDurationProperty(pathUsageReportingInterval, 15*time.Minute, "The time interval at which usage and metric event will be generated")
```

Comment:
in these 3 lines, isn't it supposed to be "events", not "event"?

Reply:
good call will change, also need to simplify the validate method for Sonar while I am in there